datalad_container/adapters/oci.py

```diff
@@ -0,0 +1,402 @@
+        name ("docker.io/library/busybox")
+    strip_transport : bool, optional
+        Remove Skopeo transport value ("docker://" or "docker-daemon:") from
+        the name. Unless this is true, reference should not include a
+        transport.
+
+    Returns
+    -------
+    A Reference namedtuple with .name, .tag, and .digest attributes
+    """
+    if strip_transport:
+        try:
+            reference = reference.split(":", maxsplit=1)[1]
+        except IndexError:
+            raise ValueError("Reference did not have transport: {}"
+                             .format(reference))
+        if reference.startswith("//"):
+            reference = reference[2:]
+
+    parts = reference.split("/")
+    last = parts[-1]
+    if "@" in last:
+        sep = "@"
+    elif ":" in last:
+        sep = ":"
+    else:
+        sep = None
+
+    tag = None
+    digest = None
+    if sep:
+        repo, label = last.split(sep)
+        front = "/".join(parts[:-1] + [repo])
+        if sep == "@":
+            digest = label
+        else:
+            tag = label
+    else:
+        front, tag = "/".join(parts), None
+    if normalize:
+        front = _normalize_reference(front)
+    return Reference(front, tag, digest)
+
+
+def _store_annotation(path, key, value):
+    """Set a value of image's org.datalad.container.image.source annotation.
+
+    Parameters
+    ----------
+    path : pathlib.Path
+        Image directory. It must contain only one image.
+    key, value : str
+        Key and value to store in the image's "annotations" field.
+    """
+    index = path / "index.json"
+    index_info = json.loads(index.read_text())
+    annot = index_info["manifests"][0].get("annotations", {})
+
+    annot[key] = value
+    index_info["manifests"][0]["annotations"] = annot
+    with index.open("w") as fh:
+        json.dump(index_info, fh)
+
+
+def _get_annotation(path, key):
+    """Return value for `key` in an image's annotation.
+
+    Parameters
+    ----------
+    path : pathlib.Path
+        Image directory. It must contain only one image.
+    key : str
+        Key in the image's "annotations" field.
+
+    Returns
+    -------
+    str or None
+    """
+    index = path / "index.json"
+    index_info = json.loads(index.read_text())
+    # Assume one manifest because skopeo-inspect would fail anyway otherwise.
+    return index_info["manifests"][0].get("annotations", {}).get(key)
+
+
+def save(image, path):
+    """Save an image to an OCI-compliant directory.
+
+    Parameters
+    ----------
+    image : str
+        A source image accepted by skopeo-copy
+    path : pathlib.Path
+        Directory to copy the image to
+    """
+    # Refuse to work with non-empty directory if it's not empty by letting the
+    # OSError through. Multiple images can be saved to an OCI directory, but
+    # run() and get_image_id() don't support a way to pull out a specific one.
+    try:
```

Comment:
Maybe we should detect and return a helpful err if skopeo is not installed? The first line of this function could be `_require_skopeo()`